src/loss.cpp

```diff
@@ -0,0 +1,83 @@
+#include <vector>
+
+#include <Eigen/Core>
+#include <deepworks/loss.hpp>
+#include "util/assert.hpp"
+
+using ConstMatrix = Eigen::Map<const Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+using Matrix = Eigen::Map<Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+
+using ConstVector = Eigen::Map<const Eigen::Matrix<float, 1, Eigen::Dynamic, Eigen::RowMajor>>;
+using Vector = Eigen::Map<Eigen::Matrix<float, 1, Eigen::Dynamic, Eigen::RowMajor>>;
+
+
+namespace deepworks {
+namespace loss {
+
+std::vector<int> GetTargetIndices(const ConstVector target, int batch_size, int n_classes) {
+    std::vector<int> slice(batch_size);
+    for (int i = 0; i < batch_size; ++i) {
+        slice[i] = static_cast<int>(target(0, i)) + n_classes * i;
+    }
+    return slice;
+}
+
+float CPUNLLLoss(Matrix X, const ConstVector target) {
+    int batch_size = X.rows();
+    int n_classes = X.cols();
+
+    Vector X_1d(X.data(), X.size());
+
+    std::vector<int> slice = GetTargetIndices(target, batch_size, n_classes);
+
+    float loss = -X_1d(0, slice).array().sum() / static_cast<float>(batch_size);
+    return loss;
+}
+
+void CPULog(const ConstMatrix X, Matrix LogX) {
```

Comment:
@TolyaTalamanov What do you think about moving this function to cpu kernels?

Reply:
All of them (`CPULog`, CPUNLLLoss`) should be in cpu kernels.
CrossEntropyLoss is a wrapper which is implemented a top of them.